File: linked-list-shift/LinkedListShift.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = next;
    }
}

// head => 0 -> 1 -> 2 -> 3 -> 4 -> 5
// k => 2
var linkedListShift = (head, k) => {
    let tail = head;
    let length = 1;
    while(tail.next != null) {
        length++;
        tail = tail.next;
    }

    let offset = Math.abs(k) % length;
    if(offset === 0) { return head; }
    let position = k > 0 ? length - offset : offset;
  
    let newTail = head;
    for(let i = 1; i < position; i++) {
        newTail = newTail.next;
    }
    let newHead = newTail.next;
    newTail.next = null; tail.next = head;
    return newHead;
}

exports.Node = Node;
exports.linkedListShift = linkedListShift;

File: linked-list-shift/LinkedListShift.cs
public class LinkedListShift {
    class Node {
        public int value;
        public Node next;
        public Node (int value) {
            this.value = value;
            this.next = null;
        }
    }

    public Node LinkedListShifts(Node head, int k) {
        Node tail = head;
        int length = 1;
        while(tail.next != null) {
            length++; tail = tail.next;
        }

        int offset = Math.Abs(k) % length;
        if(offset == 0) return head;
        int position = k > 0 ? length - offset : offset;

        Node newTail = head;
        for(int i = 1; i < position; i++) {
            newTail = newTail.next;
        }

        Node newHead = newTail.next;
        newTail.next = null; tail.next = head;
        return newHead;
    }
}